src/toil/job.py

```diff
@@ -1045,276 +1926,279 @@ def _collectAllSuccessors(self, visited):
-        return JobNode.fromJobGraph(jobGraph)
+                for successorID in job.description.allSuccessors():
+                    if successorID in self._registry:
+                        # We added this successor locally
+                        todo.append(self._registry[successorID])
 
     def getTopologicalOrderingOfJobs(self):
         """
         :returns: a list of jobs such that for all pairs of indices i, j for which i < j, \
         the job at index i can be run before the job at index j.
-        :rtype: list
+        
+        Only considers jobs in this job's subgraph that are newly added, not loaded from the job store.
+        
+        Ignores service jobs.
+        
+        :rtype: list[Job]
         """
+        
+        # List of Job objects in order.
         ordering = []
+        # Set of IDs of visited jobs.
         visited = set()
 
         # We need to recurse and traverse the graph without exhausting Python's
-        # stack, so we keep our own stack.
+        # stack, so we keep our own stack of Job objects
         todo = [self]
 
         while len(todo) > 0:
             job = todo[-1]
             todo.pop()
-
+            
+            logger.debug("Consider job %s", job.jobStoreID)
+            
             #Do not add the job to the ordering until all its predecessors have been
             #added to the ordering
             outstandingPredecessor = False
-            for p in job._directPredecessors:
-                if p not in visited:
+            for predJob in job._directPredecessors:
+                if predJob.jobStoreID not in visited:
+                    logger.debug("Has outstanding predecessor that will account for it: %s", predJob.jobStoreID)
                     outstandingPredecessor = True
                     break
             if outstandingPredecessor:
                 continue
 
-            if job not in visited:
-                visited.add(job)
+            if job.jobStoreID not in visited:
+                visited.add(job.jobStoreID)
                 ordering.append(job)
                 
-                for other in itertools.chain(job._followOns, job._children):
-                    # Stack up descendants so we process children and then follow-ons.
-                    # So stack up follow-ons deeper
-                    todo.append(other)
+                logger.debug("Visit job %s", job.jobStoreID)
+                
+                for otherID in itertools.chain(job.description.followOnIDs, job.description.childIDs):
+                    if otherID in self._registry:
+                        # Stack up descendants so we process children and then follow-ons.
+                        # So stack up follow-ons deeper
+                        todo.append(self._registry[otherID])
+                        logger.debug("Queue successor %s", otherID)
+                    else:
+                        logger.debug("Unregistered successor %s", otherID)
```

Comment:
These log messages might be noisy... ?  I'd consider removing them.  Though I have not experimented with this to know whether they are prohibitively noisy, after https://github.com/DataBiosphere/toil/issues/3274 , I think we should carefully consider logs that do not explicitly notify the user of either unexpected behavior, or a sort of throttled regular status message, since almost all log hunting seems to be searching for unexpected behavior.

Maybe better would be just a log message every ten minutes with the number of successors queued/unregistered in the last ten minutes and the total overall running/queued?  Though that might be worth making a separate PR/issue rather than continuing to add here.

Reply:
Oh yeah, I think I left a bunch of debugging in here that we don't really need. I'll pull it out.

I don't think handling these as events to report on really makes sense; they just tell you about Toil's traversal of the graphs of jobs you are adding, and I only added them to make sure the traversal ported over properly to the `JobDescription` class and the ID references.